src/sacn/source.c

```diff
@@ -83,22 +189,101 @@ void sacn_source_universe_config_init(SacnSourceUniverseConfig* config)
  */
 etcpal_error_t sacn_source_create(const SacnSourceConfig* config, sacn_source_t* handle)
 {
-  // If the Tick thread hasn't been started yet, start it if the config isn't manual.
+#if !SACN_SOURCE_ENABLED
+  return kEtcPalErrNotInit;
+#endif
 
-  ETCPAL_UNUSED_ARG(config);
-  ETCPAL_UNUSED_ARG(handle);
-  return kEtcPalErrNotImpl;
+  etcpal_error_t result = kEtcPalErrOk;
+
+  // Verify module initialized.
+  if (!sacn_initialized())
+    result = kEtcPalErrNotInit;
+
+  // Check for invalid arguments.
+  if (result == kEtcPalErrOk)
+  {
+    if (!config)
+      result = kEtcPalErrInvalid;
+    else if (ETCPAL_UUID_IS_NULL(&config->cid))
+      result = kEtcPalErrInvalid;
+    else if (!config->name)
+      result = kEtcPalErrInvalid;
+    else if (strlen(config->name) > (SACN_SOURCE_NAME_MAX_LEN - 1))  // Max length includes null terminator
+      result = kEtcPalErrInvalid;
+    else if (config->keep_alive_interval <= 0)
+      result = kEtcPalErrInvalid;
+    else if (!handle)
+      result = kEtcPalErrInvalid;
+  }
+
+  if (sacn_lock())
+  {
+    // If the Tick thread hasn't been started yet, start it if the config isn't manual.
+    if (result == kEtcPalErrOk)
+    {
+      if (!thread_initialized && !config->manually_process_source)
+      {
+        result = start_tick_thread();
+
+        if (result == kEtcPalErrOk)
+          thread_initialized = true;
+      }
+    }
+
+    // Allocate the source's state.
+    SacnSource* source = NULL;
+    if (result == kEtcPalErrOk)
+    {
+      source = add_sacn_source(get_next_int_handle(&source_handle_mgr, -1));
+
+      if (!source)
+        result = kEtcPalErrNoMem;
+    }
+
+    if (result == kEtcPalErrOk)
+    {
+      // Initialize the universe discovery send buffer.
```

Comment:
That was a gray area for me in terms of if that belonged in the mem module or not. But if you think it would fit there, then I can certainly move it over. Or perhaps if I create that source_state module I commented about elsewhere, I could make an init_source function in there?

Reply:
It does sound nice to have all this in one function instead of two though...

Reply:
I think it fits in the mem module if you think of it as containing the code necessary to implement allocation of the sACN structures, which includes their "constructors"

Reply:
Good point. I'll go ahead and pull this into add_sacn_source then.

Reply:
Resolved in https://github.com/ETCLabs/sACN/pull/32/commits/e73f667ab0f8dfd7376f21878fb96a3d6e6f438c.